src/main/java/com/uber/cadence/workflow/Saga.java

```diff
@@ -0,0 +1,313 @@
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package com.uber.cadence.workflow;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class Saga {
+  private final Options options;
+  private final List<Functions.Func<Promise>> compensationOps = new ArrayList<>();
+
+  public static final class Options {
+    private final boolean parallelCompensation;
+    private final boolean continueWithError;
+
+    private Options(boolean parallelCompensation, boolean continueWithError) {
+      this.parallelCompensation = parallelCompensation;
+      this.continueWithError = continueWithError;
+    }
+
+    public static final class Builder {
+      private boolean parallelCompensation = true;
+      private boolean continueWithError = true;
+
+      public Builder setParallelCompensation(boolean parallelCompensation) {
+        this.parallelCompensation = parallelCompensation;
+        return this;
+      }
+
+      public Builder setContinueWithError(boolean continueWithError) {
+        this.continueWithError = continueWithError;
+        return this;
+      }
+
+      public Options build() {
+        return new Options(parallelCompensation, continueWithError);
+      }
+    }
+  }
+
+  public static class CompensationException extends RuntimeException {
+    public CompensationException(Throwable cause) {
+      super("Exception from saga compensate", cause);
+    }
+  }
+
+  public Saga(Options options) {
+    this.options = options;
+  }
+
+  public void compensate() {
+    if (options.parallelCompensation) {
+      List<Promise> results = new ArrayList<>();
+      for (Functions.Func<Promise> f : compensationOps) {
+        results.add(f.apply());
+      }
+
+      CompensationException sagaException = null;
+      for (Promise p : results) {
+        try {
+          p.get();
+        } catch (Exception e) {
+          if (sagaException == null) {
+            sagaException = new CompensationException(e);
+          } else {
+            sagaException.addSuppressed(e);
+          }
+        }
+      }
+
+      if (sagaException != null) {
+        throw sagaException;
+      }
+    } else {
+      for (int i = compensationOps.size() - 1; i >= 0; i--) {
+        Functions.Func<Promise> f = compensationOps.get(i);
+        try {
+          Promise result = f.apply();
+          result.get();
+        } catch (Exception e) {
+          if (!options.continueWithError) {
+            throw e;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Add compensation operation for saga, which will be executed in the reverse order if {@link
```

Comment:
`in the reverse order` - is it true in case of parallel compensation?

Reply:
If it's parallel then there is no order...I'll call this out explicitly probably.

Reply:
So you are right continueWithError is mostly useful when you don't use parallel compensation. If you choose to do parallel compensation, then all the compensation operations will be fired no matter what and then you will receive exceptions back if there's any. I'll update the java docs.

Reply:
I'd just remove these words from addCompensation comments

Reply:
Yes will do.